subprojects/signing/src/main/java/org/gradle/plugins/signing/signatory/gnupg/GnupgSignatory.java

```diff
@@ -0,0 +1,118 @@
+
+import org.gradle.api.Action;
+import org.gradle.api.Project;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.plugins.signing.signatory.SignatorySupport;
+import org.gradle.process.ExecSpec;
+
+import java.io.File;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Signatory that uses the gnupg cli program.
+ */
+public class GnupgSignatory extends SignatorySupport {
+
+    private static final Logger LOG = Logging.getLogger(GnupgSignatory.class);
+
+    private final Project project;
+    private final String name;
+
+    private final String executable;
+    private final boolean useLegacyGpg;
+    private final File homeDir;
+    private final File optionsFile;
+    private final String keyName;
+    private final String passphrase;
+
+    public GnupgSignatory(Project project, String name, GnupgSettings settings) {
+        this.project = project;
+        this.name = name;
+        this.executable = settings.getExecutable();
+        this.useLegacyGpg = settings.getUseLegacyGpg();
+        this.homeDir = settings.getHomeDir();
+        this.optionsFile = settings.getOptionsFile();
+        this.keyName = settings.getKeyName();
+        this.passphrase = settings.getPassphrase();
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public void sign(final InputStream input, final OutputStream output) {
+        final List<String> arguments = buildArgumentList();
+        LOG.info("Invoking {} with arguments: {}", executable, arguments);
+        project.exec(new Action<ExecSpec>() {
+            @Override
+            public void execute(ExecSpec spec) {
+                spec.setExecutable(executable);
+                spec.setArgs(arguments);
+                spec.setStandardInput(input);
+                spec.setStandardOutput(output);
+            }
+        });
+    }
+
+    private List<String> buildArgumentList() {
+        final List<String> args = new ArrayList<String>();
+        if (homeDir != null) {
+            args.add("--homedir");
+            args.add(homeDir.getAbsolutePath());
+        }
+        if (optionsFile != null) {
+            args.add("--options");
+            args.add(optionsFile.getAbsolutePath());
+        }
+        if (keyName != null) {
+            args.add("--local-user");
+            args.add(keyName);
+        }
+        if (passphrase != null) {
+            if (useLegacyGpg) {
+                args.add("--no-use-agent");
+            } else {
+                args.add("--pinentry-mode=loopback");
+            }
+            args.add("--passphrase");
+            args.add(passphrase);
+        } else {
+            if (useLegacyGpg) {
+                args.add("--use-agent");
+            }
+        }
+        args.add("--no-tty");
+        args.add("--batch");
+        args.add("--armor");
```

Comment:
Adding `--armor` here causes invalid signature files to be written, because the consumer of the `OutputStream` in `Signatory.sign(...)` [already armors](https://github.com/cboehme/gradle/blob/a3e9387bb2456857bf13a5942a4d5f4d85a2e7ae/subprojects/signing/src/main/java/org/gradle/plugins/signing/type/DefaultSignatureTypeProvider.java#L30) the output. [This test](https://github.com/emlun/gradle/commit/1f717d6371014305fd9a75cbb67f50abe7ffbf56) reveals the bug.